// @chevrotain/regexp-to-ast/src/regexp-parser.ts
import type {
  Alternative,
  Assertion,
  Atom,
  Character,
  Disjunction,
  Group,
  GroupBackReference,
  Location,
  Quantifier,
  Range,
  RegExpFlags,
  RegExpPattern,
  Set,
  Term,
} from "../types";
import {
  addFlag,
  ASSERT_EXISTS,
  ASSERT_NEVER_REACH_HERE,
  cc,
  insertToSet,
  isCharacter,
} from "./utils.js";
import {
  digitsCharCodes,
  whitespaceCodes,
  wordCharCodes,
} from "./character-classes.js";

// consts and utilities
const hexDigitPattern = /[0-9a-fA-F]/;
const decimalPattern = /[0-9]/;
const decimalPatternNoZero = /[1-9]/;

// https://hackernoon.com/the-madness-of-parsing-real-world-javascript-regexps-d9ee336df983
// https://www.ecma-international.org/ecma-262/8.0/index.html#prod-Pattern
export class RegExpParser {
  protected idx: number = 0;
  protected input: string = "";
  protected groupIdx: number = 0;

  protected saveState() {
    return {
      idx: this.idx,
      input: this.input,
      groupIdx: this.groupIdx,
    };
  }

  protected restoreState(newState: {
    idx: number;
    input: string;
    groupIdx: number;
  }) {
    this.idx = newState.idx;
    this.input = newState.input;
    this.groupIdx = newState.groupIdx;
  }

  public pattern(input: string): RegExpPattern {
    // parser state
    this.idx = 0;
    this.input = input;
    this.groupIdx = 0;

    this.consumeChar("/");
    const value = this.disjunction();
    this.consumeChar("/");

    const flags: RegExpFlags = {
      type: "Flags",
      loc: { begin: this.idx, end: input.length },
      global: false,
      ignoreCase: false,
      multiLine: false,
      unicode: false,
      sticky: false,
    };

    while (this.isRegExpFlag()) {
      switch (this.popChar()) {
        case "g":
          addFlag(flags, "global");
          break;
        case "i":
          addFlag(flags, "ignoreCase");
          break;
        case "m":
          addFlag(flags, "multiLine");
          break;
        case "u":
          addFlag(flags, "unicode");
          break;
        case "y":
          addFlag(flags, "sticky");
          break;
      }
    }

    if (this.idx !== this.input.length) {
      throw Error("Redundant input: " + this.input.substring(this.idx));
    }
    return {
      type: "Pattern",
      flags: flags,
      value: value,
      loc: this.loc(0),
    };
  }

  protected disjunction(): Disjunction {
    const alts = [];
    const begin = this.idx;

    alts.push(this.alternative());

    while (this.peekChar() === "|") {
      this.consumeChar("|");
      alts.push(this.alternative());
    }

    return { type: "Disjunction", value: alts, loc: this.loc(begin) };
  }

  protected alternative(): Alternative {
    const terms = [];
    const begin = this.idx;

    while (this.isTerm()) {
      terms.push(this.term());
    }

    return { type: "Alternative", value: terms, loc: this.loc(begin) };
  }

  protected term(): Term {
    if (this.isAssertion()) {
      return this.assertion();
    } else {
      return this.atom();
    }
  }

  protected assertion(): Assertion {
    const begin = this.idx;
    switch (this.popChar()) {
      case "^":
        return {
          type: "StartAnchor",
          loc: this.loc(begin),
        };
      case "$":
        return { type: "EndAnchor", loc: this.loc(begin) };
      // '\b' or '\B'
      case "\\":
        switch (this.popChar()) {
          case "b":
            return {
              type: "WordBoundary",
              loc: this.loc(begin),
            };
          case "B":
            return {
              type: "NonWordBoundary",
              loc: this.loc(begin),
            };
        }
        /* c8 ignore next */
        throw Error("Invalid Assertion Escape");
      // '(?=' or '(?!'
      case "(":
        this.consumeChar("?");

        let type:
          | "Lookahead"
          | "NegativeLookahead"
          | "Lookbehind"
          | "NegativeLookbehind"
          | undefined;
        switch (this.popChar()) {
          case "=":
            type = "Lookahead";
            break;
          case "!":
            type = "NegativeLookahead";
            break;
          case "<": {
            switch (this.popChar()) {
              case "=":
                type = "Lookbehind";
                break;
              case "!":
                type = "NegativeLookbehind";
            }
            break;
          }
        }
        ASSERT_EXISTS(type);

        const disjunction = this.disjunction();

        this.consumeChar(")");

        return {
          type: type!,
          value: disjunction,
          loc: this.loc(begin),
        };
    }
    // istanbul ignore next
    return ASSERT_NEVER_REACH_HERE();
  }

  protected quantifier(
    isBacktracking: boolean = false,
  ): Quantifier | undefined {
    let range: Partial<Quantifier> | undefined = undefined;
    const begin = this.idx;
    switch (this.popChar()) {
      case "*":
        range = {
          atLeast: 0,
          atMost: Infinity,
        };
        break;
      case "+":
        range = {
          atLeast: 1,
          atMost: Infinity,
        };
        break;
      case "?":
        range = {
          atLeast: 0,
          atMost: 1,
        };
        break;
      case "{":
        const atLeast = this.integerIncludingZero();
        switch (this.popChar()) {
          case "}":
            range = {
              atLeast: atLeast,
              atMost: atLeast,
            };
            break;
          case ",":
            let atMost;
            if (this.isDigit()) {
              atMost = this.integerIncludingZero();
              range = {
                atLeast: atLeast,
                atMost: atMost,
              };
            } else {
              range = {
                atLeast: atLeast,
                atMost: Infinity,
              };
            }
            this.consumeChar("}");
            break;
        }
        // throwing exceptions from "ASSERT_EXISTS
